File: src/python/oop/ElevatorSystem/elevator/elevator.py
from .direction import Direction
from .elevator_status import ElevatorStatus

class Elevator:
    def __init__(self, id: str, floors: list[int]):
        self._id = id
        self._floors = set(floors)
        self._queue = []
        self._delayed_queue = []
        self._status = ElevatorStatus(self._id, 1, Direction.IDLE)

    @property
    def id(self):
        return self._id

    @property
    def floors(self) -> set[int]:
        return self._floors

    def get_status(self) -> ElevatorStatus:
        return self._status

    def add_floor_request(self, floor: int) -> None:
        if floor in self._queue or floor not in self.floors:
            return

        if (self._status.direction == Direction.UP and self._status.current_floor > floor) \
           or (self._status.direction == Direction.DOWN and self._status.current_floor < floor):
            return

        self._queue.append(floor)
        if self._status.current_floor > floor:
            self._queue.sort(reverse=True)
        else:
            self._queue.sort()

    def add_system_request(self, floor: int, direction: Direction) -> None:
        if floor in self._queue:
            return

        if self._status.direction != direction \
            or (self._status.direction == Direction.UP and self._status.current_floor > floor) \
            or (self._status.direction == Direction.DOWN and self._status.current_floor < floor):
            if floor not in self._delayed_queue:
                self._delayed_queue.append(floor)
            return

        self._queue.append(floor)

        if self._status.current_floor > floor:
            self._queue.sort(reverse=True)
        else:
            self._queue.sort()

    def _go_one_floor(self):
        if len(self._queue) <= 0:
            self._status = ElevatorStatus(self._id, self._status.current_floor, Direction.IDLE)
            return

        next_floor = self._queue[0]
        if self._status.current_floor > next_floor:
            self._status = ElevatorStatus(self._id, self._status.current_floor - 1, Direction.DOWN)
        elif self._status.current_floor < next_floor:
            self._status = ElevatorStatus(self._id, self._status.current_floor + 1, Direction.UP)
        else:
            self._queue = self._queue[1:]

            if len(self._queue) > 0:
                return

            self._status = ElevatorStatus(self._id, self._status.current_floor, Direction.IDLE)

            if len(self._delayed_queue) <= 0:
                return

            floor = self._delayed_queue[0]
            self._delayed_queue = self._delayed_queue[1:]

            self._queue.append(floor)
            self._status = ElevatorStatus(self._id, self._status.current_floor, Direction.DOWN if self._status.current_floor > floor else Direction.UP)

File: src/go/oop/ElevatorSystem/internal/elevator.go
package internal

import "sort"

type Direction int

const (
	Idle Direction = iota
	Up
	Down
)

type Elevator struct {
	Id     string
	Floors map[int]bool
	Panel  *ElevatorPanel
	status ElevatorStatus

	requestMap    map[int]bool
	requestQueue  []int
	deferredQueue map[int]bool
}

type ElevatorStatus struct {
	ElevatorId   string
	CurrentFloor int
	Direction    Direction
}

func ElevatorCar(id string, floors []int) Elevator {
	sort.Slice(floors, func(i, j int) bool {
		return floors[i] < floors[j]
	})

	buttons := map[int]bool{}

	for _, floor := range floors {
		buttons[floor] = true
	}

	elevator := Elevator{
		Id:     id,
		Floors: buttons,
		status: ElevatorStatus{
			ElevatorId:   id,
			CurrentFloor: 1,
			Direction:    Idle,
		},
	}

	panel := ElevatorCarPanel(elevator)
	elevator.Panel = &panel

	return elevator
}

func (e *Elevator) GetStatus() ElevatorStatus {
	return e.status
}

func (e *Elevator) UserRequestedFloor(floor int) {
	if _, ok := e.Floors[floor]; !ok {
		return
	}

	if _, ok := e.requestMap[floor]; ok {
		return
	}

	if (e.status.Direction == Up && e.status.CurrentFloor > floor) ||
		(e.status.Direction == Down && e.status.CurrentFloor < floor) {
		return
	}

	e.requestMap[floor] = true
	e.requestQueue = append(e.requestQueue, floor)
	sort.Slice(e.requestQueue, func(i, j int) bool {
		if e.status.CurrentFloor > floor {
			return e.requestQueue[i] < e.requestQueue[j]
		}

		return e.requestQueue[i] > e.requestQueue[j]
	})
}

func (e *Elevator) SystemRequestedFloor(floor int, direction Direction) {
	if _, ok := e.Floors[floor]; !ok {
		return
	}

	if _, ok := e.requestMap[floor]; ok {
		return
	}

	if e.status.Direction == direction ||
		((e.status.Direction == Up && e.status.CurrentFloor > floor) ||
			(e.status.Direction == Down && e.status.CurrentFloor < floor)) {
		if _, ok := e.deferredQueue[floor]; !ok {
			e.deferredQueue[floor] = true
		}
		return
	}

	e.requestMap[floor] = true
	e.requestQueue = append(e.requestQueue, floor)
	sort.Slice(e.requestQueue, func(i, j int) bool {
		if e.status.CurrentFloor > floor {
			return e.requestQueue[i] < e.requestQueue[j]
		}

		return e.requestQueue[i] > e.requestQueue[j]
	})
}
